mixins/anchor-manager.js

```diff
@@ -0,0 +1,169 @@
+// `handleNavigateToAnchor` method is called; when it needs to
+// be navigated from, its `handleNavigateFromAnchor` method is called.
+var AnchorManager = function() {
+  this.anchorMap = {};
+  this.initialized = false;
+  this.currentAnchor = null;
+  this.handleHashChange = this.handleHashChange.bind(this);
+};
+
+AnchorManager.prototype = {
+  register: function(component) {
+    var anchorId = component.props.anchorId;
+    if (process.env.NODE_ENV !== 'production' &&
+        anchorId in this.anchorMap) {
+      console.warn('Anchor for #' + anchorId + ' already registered.');
+    }
+    this.anchorMap[anchorId] = component;
+    if (!this.initialized) {
+      this.initialize();
+      this.initialized = true;
+    }
+    this.handleHashChange();
+    return component;
+  },
+  unregister: function(component) {
+    var anchorId = component.props.anchorId;
+    if (process.env.NODE_ENV !== 'production') {
+      if (!(anchorId in this.anchorMap)) {
+        console.warn('Anchor for #' + anchorId + ' does not exist.');
+      } else if (this.anchorMap[anchorId] !== component) {
+        console.warn('Anchor mismatch for #' + anchorId + '.');
+      }
+    }
+    if (this.currentAnchor === component) {
+      this.currentAnchor = null;
+    }
+    delete this.anchorMap[anchorId];
+  },
+  initialize: function() {
+    window.addEventListener('hashchange', this.handleHashChange);
+  },
+  getHash: function() {
+    return window.location.hash.slice(1);
+  },
+  handleHashChange: function() {
+    var newAnchor = this.anchorMap[this.getHash()];
+
+    if (this.currentAnchor) {
+      if (newAnchor === this.currentAnchor) {
+        return;
+      } else {
+        this.currentAnchor.handleNavigateFromAnchor();
+        this.currentAnchor = null;
+      }
+    }
+    if (newAnchor) {
+      this.currentAnchor = newAnchor;
+      this.currentAnchor.handleNavigateToAnchor();
+    }
+  }
+};
+
+var defaultAnchorManager = new AnchorManager();
+
+var AnchorManagerMixin = {
+  propTypes: {
+    anchorId: React.PropTypes.string,
+    anchorManager: React.PropTypes.object
+  },
+  getDefaultProps: function() {
+    return {
+      anchorManager: defaultAnchorManager
+    };
+  },
+  getInitialState: function() {
+    return {
+      attractAttentionToAnchor: false
+    };
+  },
+  componentDidMount: function() {
+    if (this.props.anchorId) {
+      this.props.anchorManager.register(this);
+    }
+  },
+  componentWillUnmount: function() {
+    if (this.props.anchorId) {
+      this.props.anchorManager.unregister(this);
+    }
+    this.cancelAttractAttentionToAnchor();
+  },
+  handleNavigateToAnchor: function() {
+    this.attractAttentionToAnchor();
+  },
+  handleNavigateFromAnchor: function() {
+    this.cancelAttractAttentionToAnchor();
+  },
+  attractAttentionToAnchor: function() {
+    this.setState({
+      attractAttentionToAnchor: true
+    });
+    window.clearTimeout(this.attractTimeout);
+    this.attractTimeout = window.setTimeout(
```

Comment:
I'm using timeouts here instead of relying on CSS transitions for the same reason we decided to use `TimeoutTransitionGroup` in #745. Basically, I don't trust CSS transition events and I think we should just use timeouts to be safe. :stuck_out_tongue: